File: core/src/nl/t64/game/rpg/subjects/LootSubject.java
package nl.t64.game.rpg.subjects;

import java.util.ArrayList;
import java.util.List;


public class LootSubject {

    private final List<LootObserver> observers = new ArrayList<>();

    public void addObserver(LootObserver observer) {
        observers.add(observer);
    }

    public void removeObserver(LootObserver observer) {
        observers.remove(observer);
    }

    public void removeAllObservers() {
        observers.clear();
    }

    public void notifyLootTaken() {
        observers.forEach(LootObserver::onNotifyLootTaken);
    }

    public void notifyRewardTaken() {
        observers.forEach(LootObserver::onNotifyRewardTaken);
    }

    public void notifyReceiveTaken() {
        observers.forEach(LootObserver::onNotifyReceiveTaken);
    }

}

File: core/src/nl/t64/game/rpg/subjects/LootSubject.kt
package nl.t64.game.rpg.subjects


class LootSubject {

    private val observers: MutableList<LootObserver> = ArrayList()

    fun addObserver(observer: LootObserver) {
        observers.add(observer)
    }

    fun removeObserver(observer: LootObserver) {
        observers.remove(observer)
    }

    fun removeAllObservers() {
        observers.clear()
    }

    fun notifyLootTaken() {
        observers.forEach { it.onNotifyLootTaken() }
    }

    fun notifyRewardTaken() {
        observers.forEach { it.onNotifyRewardTaken() }
    }

    fun notifyReceiveTaken() {
        observers.forEach { it.onNotifyReceiveTaken() }
    }

}
